torchaudio/models/deepspeech.py

```diff
@@ -0,0 +1,94 @@
+import torch
+import torch.nn as nn
+
+__all__ = ["DeepSpeech"]
+
+
+class FullyConnected(nn.Module):
+    """
+    Args:
+        in_features: Number of input features
+        hidden_size: Internal hidden unit size.
+    """
+
+    def __init__(self,
+                 in_features: int,
+                 hidden_size: int,
+                 dropout: float,
+                 relu_max_clip: int = 20) -> None:
+        super(FullyConnected, self).__init__()
+        self.fc = nn.Linear(in_features, hidden_size, bias=True)
+        self.nonlinearity = nn.Sequential(*[
+            nn.ReLU(),
+            nn.Hardtanh(0, relu_max_clip)
+        ])
+        if dropout:
+            self.nonlinearity = nn.Sequential(*[
+                self.nonlinearity,
+                nn.Dropout(dropout)
+            ])
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = self.fc(x)
+        x = self.nonlinearity(x)
+        return x
+
+
+class DeepSpeech(nn.Module):
+    """
+    DeepSpeech model architecture from
+    `"Deep Speech: Scaling up end-to-end speech recognition"`
+    <https://arxiv.org/abs/1412.5567> paper.
+
+    Args:
+        in_features: Number of input features
+        hidden_size: Internal hidden unit size.
+        num_classes: Number of output classes
+    """
+
+    def __init__(self,
+                 in_features: int,
+                 hidden_size: int = 2048,
+                 num_classes: int = 40,
+                 dropout: float = 0.0) -> None:
+        super(DeepSpeech, self).__init__()
+        self.hidden_size = hidden_size
+        self.fc1 = FullyConnected(in_features, hidden_size, dropout)
+        self.fc2 = FullyConnected(hidden_size, hidden_size, dropout)
+        self.fc3 = FullyConnected(hidden_size, hidden_size, dropout)
+        self.bi_rnn = nn.RNN(
+            hidden_size, hidden_size, num_layers=1, nonlinearity='relu', bidirectional=True)
+        self.nonlinearity = nn.ReLU()
+        self.fc4 = FullyConnected(hidden_size, hidden_size, dropout)
+        self.out = nn.Sequential(*[
+            nn.Linear(hidden_size, num_classes),
+            nn.LogSoftmax(dim=2)
+        ])
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        Args:
+            x (torch.Tensor): Tensor of dimension (batch_size, num_channels, input_length, num_features).
+        Returns:
+            Tensor: Predictor tensor of dimension (input_length, batch_size, number_of_classes).
```

Comment:
Thanks for updating the docstring :) as mentioned in [comment above](https://github.com/pytorch/audio/pull/1399/files#r617686286), we want to keep batch first for the output too, e.g. [wav2letter](https://github.com/pytorch/audio/blob/master/torchaudio/models/wav2letter.py#L70), [conv_tasnet](https://github.com/pytorch/audio/blob/master/torchaudio/models/conv_tasnet.py#L300), [wavernn](https://github.com/pytorch/audio/blob/master/torchaudio/models/wavernn.py#L284).

```suggestion
            Tensor: Predictor tensor of dimension (batch_size, input_length, number_of_classes).
```

Making this change might not be ideal and means the pipeline (1) needs [another transpose](https://github.com/vincentqb/stt_models/blob/dataparallel/main.py#L161) to be compatible with CTC, (2) this follows the RNN convention though `batch_first` could be used, (3) might get a performance hit (though this should be measured). Thoughts?

Reply:
All suggestions look reasonable and I have already addressed your comments, thank you for sharing it with me.

However, I'm still don't follow your 3rd point. Could you please elaborate it a bit, why keeping `batch_first` can bring a performance hit?
In the mentioned example you are using simple `DataParallel` and here is my understanding of how it works (on 2 devices for simplicity):
- the model is copied into two devices
- dataloader splits a batch into 2 chunks
- then it feeds each model by each chunk during forward pass
- then, using the outputs of models, each device calculates ctc loss, the loss values are copied into single tensor (on RAM) and reduction is performed on cpu.
- on a backward pass, the gradients are computed on 2 devices and then they are 'all-reduce'd on cpu.

If I understood the process correctly, what's the intuition of keeping `batch_first` in the model's output?

Reply:
In summary: let's have batch first for the output too for readability and consistency with the other models :)

> Could you please elaborate it a bit, why keeping `batch_first` can bring a performance hit?

The question about performance that I was raising was whether the forward method in `RNN` is "faster" when batch is first or not. For instance, this small user test [here](https://stackoverflow.com/questions/63822152/pytorch-rnn-is-more-efficient-with-batch-first-false) does not seem to find a significant difference in speed. Based on this small test, let's not worry about this here then.

> In the mentioned example you are using simple `DataParallel` and here is my understanding of how it works (on 2 devices for simplicity):

DataParallel applies only to the computation of the model:
* the model is copied to the two devices
* the batch is split in to along the first dimension and sent to each devices
* the forward is computed on each device
* the output tensors are concatenated along the first dimension
* the loss function is then computed on that new tensor

Reply:
Thank you for the detailed explanation.  Let me know if anything is wrong.